Add JenkinsMasterEvent

dalvizu opened the pull request:
JenkinsMasterEvent describes events collected every 60s from
the Jenkins master: number of online agents accepting tasks, and
whether the master is in quiet down mode or not

Fixes #11

Commit:
Add JenkinsMasterEvent

JenkinsMasterEvent describes events collected every 60s from
the Jenkins master: number of online agents accepting tasks, and
whether the master is in quiet down mode or not

Commit:
fix checkstyle issue

Commit:
add jenkins master, labels as info to event

sdewitt-newrelic commented on pom.xml:
```diff
@@ -91,6 +91,30 @@
           </execution>
         </executions>
       </plugin>
+      <plugin>
```
Hi @dalvizu, I believe I originally looked at including findbugs.  However, I'm not sure the CC license is an acceptable license for complying with New Relic's open source licensing policies.  I believe that is why I did not include it.  Also, it doesn't seem relevant to this PR.  If possible can we remove this and push it to another PR to investigate if it is acceptable?

dalvizu replied:
Sorry for not adding a note -- I added this configuration to disable the plugin. However, the plugin is already included in this build. It looks [like this is in the parent plugin](https://github.com/jenkinsci/plugin-pom/blob/plugin-3.4/pom.xml#L557).

The failures are unrelated to my change, too.

If you run `mvn help:effective-pom` you can see the definition.

If you are not seeing this behavior, could it be because of some environment configuration disabling the same?

dalvizu replied:
It looks like I can disable this with a property instead. I'll push that fix.

dalvizu replied:
It looks like upstream has an updated plugin parent pom, and this parent now uses 'spotbugs'. It found bugs unrelated to my change so I've added the spotsbugs.skip property to allow `mvn clean install` to pass

sdewitt-newrelic commented on pom.xml:
```diff
@@ -15,7 +15,7 @@
 	<groupId>com.newrelic.jenkins</groupId>
 	<artifactId>new-relic</artifactId>
-	<version>1.0.1</version>
+	<version>1.0.2-SNAPSHOT</version>
```
@dalvizu as we don't push this plugin to any maven repos, and because integrating this PR would create an actual release, please remove the `SNAPSHOT` designation.

dalvizu replied:
Will remove - I'm not familiar with your release process so was trying to be defensive about not clobbering my change with something else. I usually use maven release plugin which encourages this.

sdewitt-newrelic commented on src/main/java/com/newrelic/experts/client/model/JenkinsMasterEvent.java:
```diff
@@ -0,0 +1,45 @@
+ * </p>
+ * @author Dan Alvizu (alvizu@gmail.com)
+ */
+public class JenkinsMasterEvent
```
@dalvizu Having a sub-class here is more elegant but also seems unnecessary. The current pattern we follow is simply creating a new `Event` in the `record*` method and setting the attributes prior to being recorded.  I can see the need to have specific `*Event` classes for polymorphic reasons and if the event had a number of attributes that made the constructor require a larger number of arguments.  Where we would only need 2 here, I would prefer to just stick with the current way we do this for the other events.  If we _do_ go with separate events, I'd prefer to do it consistently so that _each_ of the events produced has it's own sub-class.

dalvizu replied:
I'll remove this class. My main reason was for having a place to encapsulate details about NewRelic (needing to pass integers as floats) and document uncommon Jenkins features (what 'quiet down mode' is)

I still think that's valuable but I definitely see the value in consistency.

sdewitt-newrelic commented on src/main/java/com/newrelic/experts/jenkins/EventRecorderImpl.java:
```diff
@@ -39,6 +40,8 @@
   private static final String CLASS_NAME = EventRecorderImpl.class.getName();
   private static final Logger LOGGER = Logger.getLogger(CLASS_NAME);
+  private final boolean isLoggingTrace = LOGGER.isLoggable(Level.FINE);
+  private final boolean isLoggingDebug = LOGGER.isLoggable(Level.FINEST);
```
@dalvizu The `EventRecordImpl` is bound as a singleton object into the DI container.  If we hold onto trace flags like this at the instance level, changes to log levels would not be picked up.  It would always use the log level that was in effect when the instance was initialized.  Please undo these changes and leave the original implementation.

dalvizu replied:
Done, good catch

sdewitt-newrelic commented on src/main/java/com/newrelic/experts/jenkins/EventRecorderImpl.java:
```diff
@@ -292,4 +291,31 @@ public void setCustomAttributes(
+  @Override
+  public void recordJenkinsMasterEvent(JenkinsMasterEvent jenkinsMasterEvent) {
+    final String methodName = "recordJenkinsMasterEvent";
+
```
@dalvizu Please add the flags here as in the other methods.

dalvizu replied:
Done

sdewitt-newrelic commented on src/main/java/com/newrelic/experts/jenkins/EventRecorderImpl.java:
```diff
@@ -292,4 +291,31 @@ public void setCustomAttributes(
   }
+
+  @Override
+  public void recordJenkinsMasterEvent(JenkinsMasterEvent jenkinsMasterEvent) {
```
@dalvizu per above can we either stick with just passing in the two parameters here (`isQuietDownMode` and `agentConnectedCount`)?

dalvizu replied:
Done

sdewitt-newrelic commented on src/main/java/com/newrelic/experts/jenkins/extensions/EventCollectorWork.java:
```diff
@@ -77,49 +77,59 @@ public long getRecurrencePeriod() {
-    final String methodName = "doRun";
-    final boolean isLoggingTrace = LOGGER.isLoggable(Level.FINE);
-    final boolean isLoggingDebug = LOGGER.isLoggable(Level.FINEST);
+    synchronized (this.recorder) {
```
@dalvizu good catch on the need to synchronize access to the event recorder.  However, this critical section is overly large.  I would prefer to modify the `EventRecorderImpl` by adding a single `synchronized private void recordEvent(Event event)` method that all the other `record*` methods call after they have prepared the event.  This would keep the critical section as small as possible.  It also means we do not need to push the need to synchronize out to all the clients of the recorder.

dalvizu replied:
Please review new changes - I propose to accomplish this now with a `Collection#synchronizedList()`, which should accomplish the goals of now decoupling clients to any synchronization, and eliminating the large `synchronized` blocks.

The concerning race condition I care about: an event might be added while the `recorder.popEvents()` method is called, which may result in events being dropped. This required synchronizing on the synchronized list object (https://imgflip.com/s/meme/Yo-Dawg-Heard-You.jpg) as this is equivalent to iterating on it: https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#synchronizedList-java.util.List-

dalvizu replied:
Please review new changes. While re-implementing per your comments, I chose to instead accomplish this with a Synchornized List data structure.

This also required a `synchronized` on the synchronized list (https://imgflip.com/s/meme/Yo-Dawg-Heard-You.jpg) during `recorder.popEvents()` method, as it is iterating on the list https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#synchronizedList-java.util.List- 

The only concerning boundary condition is an event being added while the collector runs, and losing events. I believe this is covered.

sdewitt-newrelic commented on src/main/java/com/newrelic/experts/jenkins/extensions/EventCollectorWork.java:
```diff
@@ -77,49 +77,59 @@ public long getRecurrencePeriod() {
+      if (isLoggingTrace)
+      {
+        LOGGER.exiting(CLASS_NAME, methodName);
+      }
```
@dalvizu All the formatting changes above should go away once we move the `synchronized` block down into the recorder.  Also, please abide by the `checkstyle` rules.  Opening braces should go on the same line as the `if`.

dalvizu replied:
Done. I've added the `LeftCurly` rule to checkstyle to enforce this.

sdewitt-newrelic commented on src/main/java/com/newrelic/experts/jenkins/extensions/JenkinsMasterEventPeriodicWork.java:
```diff
@@ -0,0 +1,100 @@
+  
+  @Override
+  protected void doRun() throws Exception {
+    synchronized (this.recorder) {
```
@dalvizu given the above changes I requested to push the `synchronized` section down to the recorder, you should be able to remove it here.

dalvizu replied:
Done

sdewitt-newrelic commented on src/main/java/com/newrelic/experts/jenkins/extensions/JenkinsMasterEventPeriodicWork.java:
```diff
@@ -0,0 +1,100 @@
+        LOGGER.entering(CLASS_NAME, methodName);
+      }
+
+      JenkinsMasterEvent jenkinsMasterEvent = new JenkinsMasterEvent();
```
@dalvizu You won't need to create the event anymore.  Just pass the two variables into your modified `record*` method.

dalvizu replied:
Done

sdewitt-newrelic commented on src/test/java/com/newrelic/experts/JenkinsMasterEventTest.java:
```diff
@@ -0,0 +1,31 @@
+ * @author dalvizu
+ */
+@Test(groups = { "unit" })
+public class JenkinsMasterEventTest
```
@dalvizu No need for this test if we remove the `JenkinsMasterEvent`.

dalvizu replied:
Done

sdewitt-newrelic commented on src/main/java/com/newrelic/experts/jenkins/extensions/JenkinsMasterEventPeriodicWork.java:
```diff
@@ -0,0 +1,100 @@
+ * @author Dan Alvizu (alvizu@gmail.com)
+ */
+@Extension
+public class JenkinsMasterEventPeriodicWork
```
@dalvizu to stay consistent with the naming of `EventCollectorWork`, can we take the `Periodic` out of this class name please?

dalvizu replied:
Done

sdewitt-newrelic commented on src/test/java/com/newrelic/experts/jenkins/EventRecorderImplTest.java:
```diff
@@ -0,0 +1,60 @@
+ * @author dalvizu
+ */
+@Test(groups = { "unit" })
+public class EventRecorderImplTest {
```
@dalvizu appreciate you adding this test.  We can test the `recordJenkinsMasterEvent()` here as well.  Also, on the current `master`, I had to migrate over to JUnit but the changes should be minimal.

dalvizu replied:
I've ported this test to JUnit and added mocks around label and hostname methods we query Jenkins on

sdewitt-newrelic commented:
@dalvizu I actually have another recent request to add information about executors.  I was thinking of combining this with your requests and creating a `JenkinsSystemEvent` with your attributes as well as with those in my other request.

sdewitt-newrelic commented:
@dalvizu Can you have a look at #36?  I think this integrates your code and addresses some of the concerns I had above.  It also adds other system level data such as node/queue data.

Commit:
Updates per pull request comments

dalvizu reviewed (commented):
Thanks for the review, updated based on comments

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
port to junit and fix tests

sdewitt-newrelic commented:
@dalvizu in case my comment above got lost in the shuffle, can you have a look at #36 ?  I had a request to add node and queue data so what I added the quite mode and connected agent count to that PR as well and refactored based on our discussions above.  Can you have a look for me?

dalvizu commented:
> @dalvizu Can you have a look at #36? I think this integrates your code and addresses some of the concerns I had above. It also adds other system level data such as node/queue data.

Ah, wish I held off for a bit :)

I'm definitely a fan of that approach. Of the changes I made today, a few can be integrated to #36:

* Disabling findbugs / spotbugs
* Add LeftCurly checkstyle rules
* mockito-inline addition to allow mocking of Jenkins calls required to test setHostname / setLabel

I can do that after you've merged #36, or you're welcome to just bring them in there. Either way I think we can close this.

sdewitt-newrelic commented:
@dalvizu can you add that comment to #36 so we have it there for the record as well?  Then if you don't mind giving it a full review and adding any comments, I'd appreciate it.  Specifically can you see if it addresses the synchronization issue you mention above?

sdewitt-newrelic commented:
Closing in favor of #36.
